### PR TITLE
feat(changelogs): defer retrieval if not required for templates

### PR DESCRIPTION
**Pull request by viceice**
## Changes
- defer changelogs again
- skip changelogs if not onboarded
- only fetch changelogs early when required by templates

## Context
- #6337
- #6346
- #17063

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/workers/repository/updates/branchify.spec.ts [hunk 1]
```diff
@@ -153,5 +153,38 @@ describe('workers/repository/updates/branchify', () => {
...
+          depName: 'bar',
+          branchName: 'bar-{{version}}',
+          prTitle: 'some-title',
+          version: '1.1.0',
+          groupName: 'My Group',
+          group: { branchName: 'renovate/my-group' },
+        },
+      ]);
+      const res = await branchifyUpgrades(config, {});
+      expect(embedChangelogs).toHaveBeenCalled();
```
Comment (zharinov) [inline lib/workers/repository/updates/branchify.spec.ts:186]: Maybe make number of calls more clear?

```suggestion
      expect(embedChangelogs).toHaveBeenCalledOnce();
```

File: lib/workers/repository/changelog/index.ts [hunk 2]
```diff
@@ -1,21 +1,31 @@
...
-  for (const branchName of Object.keys(branchUpgrades)) {
-    for (const upgrade of branchUpgrades[branchName]) {
-      upgrades.push(upgrade);
+  await pMap(branches, embedChangelog, { concurrency: 10 });
+}
+
+export function needsChangelogs(upgrade: BranchUpgradeConfig): boolean {
+  for (const field of exposedConfigOptions) {
+    // fields set by `getChangeLogJSON`
+    if (containsTemplates(upgrade[field], ['logJSON', 'releases'])) {
```
Comment (zharinov) [inline lib/workers/repository/changelog/index.ts:26]: What if make `['logJSON', 'releases']` the part of `containsTemplates()`?

**Review by zharinov — CHANGES_REQUESTED** [review]

File: lib/workers/repository/changelog/index.ts (hunk 2 above)
Comment (viceice) [reply lib/workers/repository/changelog/index.ts:26]: @rarkins  WDYT?

Would require renaming too, as it's only testing for specific templates

**Review by zharinov — DISMISSED** [review]
Otherwise LGTM

File: lib/workers/repository/changelog/index.ts (hunk 2 above)
Comment (rarkins) [reply lib/workers/repository/changelog/index.ts:26]: 🤷

Comment (viceice) [reply lib/workers/repository/changelog/index.ts:26]: so let's leave it here, as it should be removed in a major bump

**Comment by viceice** [comment]
can't resolve from mobile.

@zharinov @JamieMagee can you fix the conflict?

**Comment by zharinov** [comment]
I can't push to your branch 🤷‍♂️

**Review by rarkins — DISMISSED** [review]

**Review by rarkins — DISMISSED** [review]

File: lib/util/template/index.spec.ts [hunk 3]
```diff
@@ -123,4 +127,34 @@ describe('util/template/index', () => {
       expect(obj[forbiddenField]).toBeUndefined();
     });
   });
+  describe('containsTemplate', () => {
```
Comment (viceice) [inline lib/util/template/index.spec.ts:130]:
```suggestion

  describe('containsTemplate', () => {
```

**Review by rarkins — APPROVED** [review]

**Pull request merged**